File: army.py
# kb 2-7-2025
# simulator to experiment with Lychgate RPG combat scenarios
# combat unit Army container class

import comUn

class Army:
    def __init__(self, lsUnits):
        self.units = lsUnits
        self.sortSelf()
    
    def unitIndexer(self, n):
        #uses initiative (ini) as index
        return n.ini
    
    def sortSelf(self):
        units = self.units
        units.sort(key = self.unitIndexer)
        units.reverse()
        self.units = units
        return units 
    
    def add(self, inUnit):
        units = self.units
        units.append(inUnit)
        self.units = units
        self.sortSelf()
        pl = len(units)
        return pl
        
    def report(self):
        units = self.units
        for u in units:
            print("------------")
            print(u.name)
            print("------------")
            print("Hit Points: ", u.hp)
            if (u.alive):
                print("Status: Alive")
            else:
                print("Status: Deceased")
            print("Net Accuracy: ", u.hitChance)
            print("Damage: ", u.dmg)
            print("Armor: ", u.ac)
            print("Net Evasion: ", u.dodge)
            print("Initiative: ", u.ini)
            print("")
            

    def top(self,i):
       ## print("method called.", i)
        #access and remove first element
        if (len(self.units)>=i+1):
            ##print("length correct.")
            top = self.units[i]
            if (top.ready):
                ##print("returning top of pack:", top.name)
                return top
            else:
                ##print("going deeper...")
                #otherwise recursive call to find the next in line who is ready
                # ready = no action this round
                return self.top(i+1)
        else: 
            #if there's noone left, return a blank unit with idr=0
            #this terminates the round if both armies have idr=0 as top
            return comUn.comUn('null',0,0,0,0,0,0,0)
    
    def topOne(self):
        return self.top(0)    
    def getCount(self):
        return len(self.units)
        
    def getTopTarget(self):
        return self.units[0]

File: dotnet/Army.cs
using Lychgate;

//todo: import comUn class

public class Army
{
	//define instance variables
	//list of units; List predefined to hold type ComUn (combatUnit)
	private List<ComUn> unitList = new List<ComUn>();
	public Army(List<ComUn> inListUnits)
	{
		//take in list of units
		this.unitList = inListUnits;
		this.SortSelf();
		return;
	}

	public bool SortSelf()
	{

		//make local version
		List<ComUn> units = this.unitList;
        //sort using initative of ComUn as key
        //comparison delegate see: https://youtu.be/uRaO8HLoHAs?t=205
        units.Sort((unit1, unit2) => unit1.Ini.CompareTo(unit2.Ini));

		//reverse list since we want to sort descending instead of asc
		units.Reverse();
        //swap instance list with sorted list
        this.unitList = units;
		return true;
	}

	public int GetCount()
	{
		int len = this.unitList.Count;
		return len;

	}

	public int Add(ComUn inUnit)
	{
		//adds a unit, sorts itself, then returns the new length
		this.unitList.Add(inUnit);
		this.SortSelf();
		return this.GetCount();
	}

	public List<String> Report()
	{
		//define a bus of strings; a list of strings that comprise a full report message.
		List<String> rptStrBus = new List<String>();

		//header
		const String margin = "------------";
		String line = "";
		//iterate through unit list
		List<ComUn> units = this.unitList;
		foreach (ComUn u in units)
		{
			
			rptStrBus.Append(margin);
			line = u.Name + " id: #" + u.Idr;
			rptStrBus.Append(line);
            rptStrBus.Append(margin);
			line = "Hit Points: " + u.Hp;
            rptStrBus.Append(line);
            if (u.Alive)
			{
                line = "Status: Alive";

            } else
			{
				line = "Status: Dead";
			}
			rptStrBus.Append(line);
			line = "Accuracy: " + u.Accuracy;
            rptStrBus.Append(line);
			line = "Damage: " + u.Dmg;
			rptStrBus.Append(line);
			line = "Evasion Chance: " + u.Dodge;
			rptStrBus.Append(line);
			line = "Initiative: " + u.Ini;
			rptStrBus.Append(line);
			line = "";
            rptStrBus.Append(line);

        }

        return rptStrBus;
	}

	private ComUn Top(int index)
	{
		//recursive function to find the top ready unit
		ComUn rtnUnit;
		//check if the index is out of range
		if (this.unitList.Count >= index + 1)
		{
			//inspect the unit of the current index
			rtnUnit = this.unitList[index];
			//check if that unit is ready
			if (rtnUnit.Ready)
			{
				//if the unit is ready, then they are the top ready unit.
				return rtnUnit;

			} 
			else
			{
				//otherwise we need to go to the next lowest initiative.
				//because the Army class is self-sorting, we know the next index will be the next highest initative.
				//so we increment our current index then call this method on itseld again.
				return this.Top(index + 1);
			}
		}
		else
		{
			//if we reach this section, it means there's noone left in the army.
			//to signify this, we return a unit with a NULL name and all 0s
			return new ComUn("null", 0, 0, 0, 0, 0, 0, 0);
			//this terminates the round sequence in the Battle class.
		}

	}
    public ComUn TopOne()
    {
        ComUn payload = this.Top(0);
		return payload;
	}
	public ComUn GetTopTarget(int index)
	{
        // this is the SAME LOGIC as this.Top() with the difference being that this
        // only returns the top ALIVE unit instead of top ready unit

		ComUn rtnUnit;
        //check if the index is out of range
        if (this.unitList.Count >= index + 1)
        {
            //inspect the unit of the current index
            rtnUnit = this.unitList[index];
			//check if that unit is ready
			if (rtnUnit.Alive)
            {
                //if the unit is ready, then they are the top ready unit.
                return rtnUnit;

            }
            else
            {
                //otherwise we need to go to the next lowest initiative.
                //because the Army class is self-sorting, we know the next index will be the next highest initative.
                //so we increment our current index then call this method on itseld again.
                return this.Top(index + 1);
            }
        }
        else
        {
            //if we reach this section, it means there's noone left in the army.
            //to signify this, we return a unit with a NULL name and all 0s
            return new ComUn("null", 0, 0, 0, 0, 0, 0, 0);
            //this terminates the round sequence in the Battle class.
        }


    }

	public ComUn GetTopTargetOne()
	{

		return GetTopTarget(0);
	}

	public void ReadyUpAll()
	{
		// in python we do this by passsing by reference and using local references
		// C# is pass-by-value so I implemented ComUn.ReadyUp() in that class
		foreach (ComUn u in this.unitList)
		{
			u.ReadyUp();
		}
	}
}

